### PR TITLE
feat: add Open Telemetry to trace fault code response Chk 1754 

### PR DESCRIPTION
**Pull request by ciuffagianluca**
Add open telemetry to trace fault code response

#### List of Changes

Added open telemetry
Added open telemetry configuration
Trace nodo error with dedicate span attribute "faultCode"

#### Motivation and Context

This feature is needed to generate dashboard in ELK for payment-request monitoring

#### How Has This Been Tested?

Test has been performed in dev env

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/openTelemetry/OpenTelemetryConfiguration.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import io.opentelemetry.api.OpenTelemetry
+import io.opentelemetry.api.trace.Tracer
+import org.springframework.context.annotation.Bean
+import org.springframework.context.annotation.Configuration
+
+@Configuration
+class OpenTelemetryConfiguration {
+
+  @Bean
+  fun agentOpenTelemetrySDKInstance(): OpenTelemetry? {
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/openTelemetry/OpenTelemetryConfiguration.kt:13]:
```suggestion
  fun agentOpenTelemetrySDKInstance(): OpenTelemetry {
```

File: src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/openTelemetry/OpenTelemetryConfiguration.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    return GlobalOpenTelemetry.get()
+  }
+
+  @Bean
+  fun openTelemetryTracer(openTelemetry: OpenTelemetry): Tracer? {
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/openTelemetry/OpenTelemetryConfiguration.kt:18]:
```suggestion
  fun openTelemetryTracer(openTelemetry: OpenTelemetry): Tracer {
```

File: src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/openTelemetry/util/OpenTelemetryUtils.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.stereotype.Component
+
+@Component
+class OpenTelemetryUtils(
+  @Autowired private val openTelemetryTracer: Tracer,
+) {
+
+  fun addSpanWithAttributes(spanName: String?, attributes: Attributes?) {
+    val span = openTelemetryTracer.spanBuilder(spanName!!).startSpan()
+    span.setAllAttributes(attributes!!)
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/openTelemetry/util/OpenTelemetryUtils.kt:15]:
```suggestion
  fun addSpanWithAttributes(spanName: String, attributes: Attributes) {
    val span = openTelemetryTracer.spanBuilder(spanName).startSpan()
    span.setAllAttributes(attributes)
```

File: src/main/kotlin/it/pagopa/ecommerce/payment/requests/services/PaymentRequestsService.kt [hunk 4]
```diff
@@ -128,6 +133,13 @@ class PaymentRequestsService(
       return@flatMap paymentRequestInfo
     }
 
+  private fun traceNodoError(faultCode: String?) {
+    faultCode?.let { code ->
+      openTelemetryUtils.addSpanWithAttributes(
+        "nodoError", Attributes.of(AttributeKey.stringKey("faultCode"), code))
+    }
+  }
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/payment/requests/services/PaymentRequestsService.kt:141]: I would let the caller handle the `null` case

```suggestion
  private fun traceNodoError(faultCode: String) {
    openTelemetryUtils.addSpanWithAttributes(
      "nodoError", Attributes.of(AttributeKey.stringKey("faultCode"), faultCode))
  }
```

File: src/main/kotlin/it/pagopa/ecommerce/payment/requests/services/PaymentRequestsService.kt [hunk 5]
```diff
@@ -128,6 +133,13 @@ class PaymentRequestsService(
       return@flatMap paymentRequestInfo
     }
 
+  private fun traceNodoError(faultCode: String?) {
+    faultCode?.let { code ->
+      openTelemetryUtils.addSpanWithAttributes(
+        "nodoError", Attributes.of(AttributeKey.stringKey("faultCode"), code))
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/payment/requests/services/PaymentRequestsService.kt:139]: Better to make this a declared constant somewhere

File: src/main/kotlin/it/pagopa/ecommerce/payment/requests/services/PaymentRequestsService.kt [hunk 6]
```diff
@@ -128,6 +136,13 @@ class PaymentRequestsService(
       return@flatMap paymentRequestInfo
     }
 
+  private fun traceVerifyNodoError(faultCode: String?) {
+    openTelemetryUtils.addErrorSpanWithAttributes(
+      VERIFY_PAYMENT_NOTICE_NODO_ERROR_SPAN_NAME,
+      Attributes.of(
+        AttributeKey.stringKey(FAULT_CODE_SPAN_KEY), Optional.ofNullable(faultCode).orElse("null")))
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/payment/requests/services/PaymentRequestsService.kt:143]:
```suggestion
        AttributeKey.stringKey(FAULT_CODE_SPAN_KEY), Optional.ofNullable(faultCode).orElse("No faultCode received")))
```
Use here the same default string used for transactions-service

File: src/test/kotlin/it/pagopa/ecommerce/payment/requests/services/PaymentRequestsServiceTests.kt [hunk 7]
```diff
@@ -190,6 +199,12 @@ class PaymentRequestsServiceTests {
       }
     /** assertions */
     assertEquals("PPT_STAZIONE_INT_PA_IRRAGGIUNGIBILE", exception.faultCode)
+    verify(openTelemetryUtils, Mockito.times(1))
+      .addErrorSpanWithAttributes(
+        any(),
```
Comment (pietro-tota) [inline src/test/kotlin/it/pagopa/ecommerce/payment/requests/services/PaymentRequestsServiceTests.kt:204]: Add check also against expected span name

File: src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/openTelemetry/util/OpenTelemetryUtils.kt [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+  /**
+   * Add an error span to the current transaction with the input arguments and name
+   *
+   * @param spanName
+   * - error span name
+   * @param attributes
+   * - error span attributes
+   */
+  fun addErrorSpanWithAttributes(spanName: String?, attributes: Attributes?) {
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/openTelemetry/util/OpenTelemetryUtils.kt:44]: See above comment

File: src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/openTelemetry/util/OpenTelemetryUtils.kt [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+  /**
+   * Add an error span to the current transaction with the input name and throwable stacktrace
+   *
+   * @param spanName
+   * - error span name
+   * @param throwable
+   * - error cause Throwable
+   */
+  fun addErrorSpanWithException(spanName: String?, throwable: Throwable?) {
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/openTelemetry/util/OpenTelemetryUtils.kt:59]: See above comment

**Review by pietro-tota — APPROVED** [review]
🚀

**Review by pietro-tota — APPROVED** [review]

**Pull request merged**